### ISSUE TITLE
Vectorization of selective masking

### ISSUE THREAD

**Issue by mhugo** [qgis/QGIS-Enhancement-Proposals#186]
# Vectorization of selective masking

**Date** 2020/05/29

**Authors**
- Bertrand Rix (@obrix)
- Hugo Mercier (@mhugo)

**Contact** infos at oslandia dot com

**maintainer** @obrix

**Version** QGIS 3.16

# Summary

The "selective masking" feature has been added to QGIS 3.12 and allows users with advanced cartography design needs to greatly improve readibility of their maps.

It is used to design areas around some labels or point markers which can "un-draw" chosen symbol layers of underlying layers.

However the current implementation forces rasterization of the whole map, which means prints and exports to vector formats (PDF, SVG) can not easily be reworked with third party publication software (Inkscape for instance).

This QEP proposes to discuss alternatives to the current implementation, so that rasterization is avoided as much as possible.

*Note* This work is proposed for the QGIS grant programme 2020 as the polishing of an existing feature.

## Proposed Solutions

No definitive technical solution has been chosen yet.

Two technical solutions are considered here:
- a first "hybrid" solution where the whole map stays vectorized, and only mask areas are rendered as bitmaps.
- a second "full vector" solution where the whole map as well as the mask parts both stay vectorized.

The "full vector" solution will be the main objective, but some implementation details are not totally fixed yet. If it happens to be too complex to implement, the "hybrid" solution will then be implemented.

### Simplified view of the current implementation

The current implementation has added a second rendering pass when selective masking is required.

During the first pass, layers are rendered each one in a separate image. Masks (label masks or masks for markers) are rendered in another image.

In a second pass, layers that have symbol layers that are masked are rendered again but with some symbol layers disabled.

Then a composition stage limits the second pass rendering to lies inside mask regions and superimpose this mask "patches" onto the image of the first pass rendering.

The composition stage uses [Composition Modes](https://doc.qt.io/qt-5/qpainter.html#composition-modes) of QPainter to merge mask parts onto the first pass image.

A final composition of all layer images is done to obtain the final image.

The fact that the rendering is done for each layer in an isolated image makes possible an optimization: layers that have no masked symbol layers are rendered only once.
However this approach comes with a price: the rasterization is forced.

### Hydrid solution

With this solution, the rendering is not necessarily separated into layers anymore. For instance, for a sequential rendering, all the layers are normally rendered to the output QPainter.

The second pass will then be rendered on a QImage, as well as a global mask image. Only symbol layers that are **not** masked are rendered during this second pass, others are disabled. Parts of the second pass image that lie inside the mask image will then be pasted over the first pass rendering. Only these mask patches will then be rasterized.

The main difference with the current implementation is that during the second pass **all** layers will be rendered again, whether or not some of their symbol layers are masked.

### Full vector solution

With this solution, the bitwise composition operations of Qt must be replaced by their "vector" equivalent in order to avoid rasterization.

The composition stage where the mask image is used to cut the rendering of the second pass should be replaced by the use of a [clip path](https://doc.qt.io/qt-5/qpainter.html#setClipPath).

The first pass rendering won't allocate QImage when not needed. During this stage, rather than computing mask **images**, **outline** of masks will be computed and stored as a [QPainterPath](https://doc.qt.io/qt-5/qpainterpath.html).

The second pass rendering will then be done over the first pass with a clip path set to the **outline** of masks.

The outline of label masks can be computed thanks to utilitary functions in Qt, especially [QPainterPathStroker](https://doc.qt.io/qt-5/qpainterpathstroker.html).

Computing the outline of masks used for markers are more difficult considering the current design. Masks on point layers are specified by the use of a ["mask marker symbol layer"](https://www.qgis.org/api/classQgsMaskMarkerSymbolLayer.html) which defines the mask image by the rendering of a sub symbol, which itself can be anything.

The mask marker symbol layer will then be redesigned and simplified so that only simple shapes that can easily be converted to QPainterPath will be allowed.

Masks with bitmap effects (outer glow effect on the alpha channel for instance) may require to compute mask images during the first pass, in addition to their outlines. For some effects, rasterization might be inevitable.

### Affected Files

- core/qgsmaprendererjob.cpp mainly

## Performance Implications

For projects with lots of layers and only some of them have masks defined, the depicted approches will more or less double the rendering time.

If no masks are defined no penalty should be involved.

The "full vector" solution may also suffer from a slight performance penalty due to computation of outline paths, compared to the current implementation.

## Votes

(required)

**Comment by nyalldawson** [comment]
Before asking for a grant for this, have you considered approaching the original funders for the follow up work?

**Comment by nyalldawson** [comment]
I should add - I'm really excited to read this. I think it's very much needed as the current approach makes the feature unusable for high quality cartographic work.

Regarding the speed: you could use the render context "force vector" flag to detect whether vector based rendering is desired and use the current faster approach when this flag isn't set. That way it would only apply when outputting to pdf/svg, so you'd limit the extra rendering time.

**Comment by nyalldawson** [comment]
> _The mask marker symbol layer will then be redesigned and simplified so that only simple shapes that can easily be converted to QPainterPath will be allowed.

I'd suggest we add a new virtual method to the market symbol layer to return a QgsGeometry representation of the marker shape. We could then buffer this for the mask, but it would also be useful for:

1. Calculating better label placements for point features, by allowing us to detect the intersection of "rays" with the symbol boundary. Currently we just use a rectangular bounds for this, which is suboptimal for circular and any non-rectangular markers

2. Allow an equivalent of ArcMap's "halo". Option for markers. These are basically the equivalent of label buffers, but drawn underneath point symbols.

**Comment by andreasneumann** [comment]
> Before asking for a grant for this, have you considered approaching the original funders for the follow up work?

Yes, we discussed this. The original funding was a concerted effort of several funders (mainly or entirely from Switzerland) and it is difficult to approach each of these funders again for follow-up work. The whole project, during implementation, turned out to be more complex than originally envisaged, also Oslandia did more work than they expected in the offer. In order to complete it in a satisfactory way we need to source different funding sources. The original funders (e.g. the Swiss QGIS user group) will also participate (e.g. for helping to solve https://github.com/qgis/QGIS/issues/34352 ), but can't shoulder everything.

Besides, it wasn't an explicit requirement in the original proposal to keep everything as vectors. This is a requirement mainly necessary for people who want to do post-processing in a vector editor.

**Comment by andreasneumann** [comment]
> I should add - I'm really excited to read this. I think it's very much needed as the current approach makes the feature unusable for high quality cartographic work.

I partially disagree on this statement. It is only necessary if you need to do follow-up editing in a vector graphics editor. If your output is a raster in a higher resolution, then this requirement isn't strictly necessary. But I agree, it would be good to see it supported.

**Comment by mhugo** [comment]
> I'd suggest we add a new virtual method to the market symbol layer to return a QgsGeometry representation of the marker shape. We could then buffer this for the mask, but it would also be useful for:
> 
>     1. Calculating better label placements for point features, by allowing us to detect the intersection of "rays" with the symbol boundary. Currently we just use a rectangular bounds for this, which is suboptimal for circular and any non-rectangular markers
> 
>     2. Allow an equivalent of ArcMap's "halo". Option for markers. These are basically the equivalent of label buffers, but drawn underneath point symbols.

Good to read there would be other applications for the geometry of markers.

But it seems not that easy to compute an outline geometry for all the different marker types. How would you do for SVGs for instance ?
I was thinking about something like a custom QPaintDevice that would output a QPainterPath (or QgsGeometry) from raw drawing primitives. Advantage: it is generic and does not require every marker class to implement a new method. But it is probably a bit of a hack ...

**Comment by nyalldawson** [comment]
> But it seems not that easy to compute an outline geometry for all the different marker types. How would you do for SVGs for instance ?

Good question :laughing: ... I've got to admit, I was thinking initially of only supporting the easy cases.

> I was thinking about something like a custom QPaintDevice that would output a QPainterPath (or QgsGeometry) from raw drawing primitives. Advantage: it is generic and does not require every marker class to implement a new method. But it is probably a bit of a hack ...

I like this approach -- it seems much more practical, and would work really well for the "halo" use case! I'd say ignore my point about label placement -- this could be better addressed using a different virtual method for marker symbol layers (something like "calculate label position offset from marker bounds") because if we do a dedicated method just for this then it would avoid the cost of converting the symbol to QgsGeometry and subsequently calculating intersections with it... we could just do everything using primitive shape calculations instead.

**Comment by obrix** [comment]
Hello ! I have started the implementation of a prototype to test the full vector solution. Here is a small advancement report :

* I reused existing painter path in QgsSimpleMarkerSymbolLayerBase to generate and aggregate the outline of masks in the render context.
* I reused existing painter path of text label in QgsTextRenderer in combination with QPainterStroke to create the outline of label masks.
* I modified the preparation of the rendering jobs and the composition of the second rendering pass to render on QPicture and compose from them if the option to force vector render is set to true. Clipping is used to remove masked part of the rendering in the composition with the aggregated QPainterPath.
As of now, this allows to get full vector output of a map with selective masking enabled through the qgis layout pdf export. [Simple test with label.pdf](https://github.com/qgis/QGIS-Enhancement-Proposals/files/5002033/testLabel.pdf)  and [Test demo.pdf](https://github.com/qgis/QGIS-Enhancement-Proposals/files/5002038/Test.demo.pdf) are two output examples with some simple demo projects of selective masking.

So far this experiment exhibit a few issues :
* Mask opacity is ignored as clip path completely cut some part of the image. I think this is possible to reimplement this pretty simply in the full vector solution by using the opacity attribute of the QPainter and repainting the masked part of the image with an opacity of the mask layer. 
  I'm not sure about the performance impact though. I also am not sure that more complex effects using gradient of alpha blending will be doable in the full vector solution.

* QSvgGenerator does not seem to handle clipping at all. Thus selective masking may have to stay fully rasterized for svg export from the layout editor. Any idea about that ?

* The phantom lines issue treated [here](https://github.com/qgis/QGIS/pull/36697) is sometimes back when visualizing the exported pdf in specific tools. The vectorized previsualisation of the map in the qgis layout tool does not exhibit this issue. Opening the pdf in Inkscape also gives a perfect result. The artefact is visible in the default pdf viewer on ubuntu and also when showing the pdf in a browser such as firefox. I think this may be linked to how those different tools handle clipping for the rendering ?

* Last small issue is that when the dpi in the layout export tool is set to something different than the fixed DPI of QPicture (96 on my machine), the exported pdf is scaled out of the frame. I think this could be fixed by applying a transformation to the drawn qpicture using a ratio between default and export dpi, but maybe there is a better way ?

The code is available [here](https://github.com/obrix/QGIS/tree/smaskingvecto_proto). The full vector implementation lives next to the old rasterized implementation. This is a WIP experimentation probably unstable and not implemented in the most efficient way as I was not familiar at all with this part of QGIS : ) I will continue to work on this to try to fix all those issues.

**Comment by andreasneumann** [comment]
> * QSvgGenerator does not seem to handle clipping at all. Thus selective masking may have to stay fully rasterized for svg export from the layout editor. Any idea about that ?

Unfortunately qt decided to implement only the SVG tiny specification. The clipPath element is only part of the SVG full 1.1 specification.

So  either QGIS is actually doing the clipping (e.g. through GEOS) before exporting the SVG, or we offer to a) either ignore selective masking in SVG export or b) rasterize it - or we get qt to implement clipPath and other useful parts of the SVG 1.1 specification.

The latter is quite unlikely unless we sponsor that development - and even then it isn't sure if they would do it.

**Comment by andreasneumann** [comment]
> * Mask opacity is ignored as clip path completely cut some part of the image. I think this is possible to reimplement this pretty simply in the full vector solution by using the opacity attribute of the QPainter and repainting the masked part of the image with the opacity of the mask layer.

Opacity wasn't part of the original requirement - only hard clips were required - someone else - who didn't help to finance (I can't remember who?) asked for that. I wouldn't mind dropping that.

Or alternatively, we would do what you suggest, and do that repainting

In general, I think performance degradation through vectorized clipping is acceptable. As I understand this, there would be a switch that would enable this - only - for export (PDF and SVG). It doesn't matter much, if that process takes slightly longer. Exporting maps is not something you do all the time.

**Comment by nyalldawson** [comment]
@obrix 
> Hello ! I have started the implementation of a prototype to test the full vector solution. Here is a small advancement report :

Great news!

> I reused existing painter path in QgsSimpleMarkerSymbolLayerBase to generate and aggregate the outline of masks in the render context.
> I reused existing painter path of text label in QgsTextRenderer in combination with QPainterStroke to create the outline of label masks.

This sounds perfect. QPainterStroke will be slow, but I think in the context of this setting then an impact on performance is perfectly acceptable.

> I modified the preparation of the rendering jobs and the composition of the second rendering pass to render on QPicture and compose from them if the option to force vector render is set to true. Clipping is used to remove masked part of the rendering in the composition with the aggregated QPainterPath.

Sounds great! One quick tip: when you set the painter clip path, make sure you use the option to intersect it with the existing clip path as these are frequently set from outside the rendering code (e.g. to clip the map to the extent of a layout map frame). You'll need to save the previous clip path and restore it after -- the QgsScopedQPainterState class is a handy utility for this.

> As of now, this allows to get full vector output of a map with selective masking enabled through the qgis layout pdf export. Simple test with label.pdf and Test demo.pdf are two output examples with some simple demo projects of selective masking.

Looking at "test.demo", it seems like there's duplicate symbols being rendered here. I think that could possible account for the phantom lines showing here...

> Mask opacity is ignored as clip path completely cut some part of the image. I think this is possible to reimplement this pretty simply in the full vector solution by using the opacity attribute of the QPainter and repainting the masked part of the image with the opacity of the mask layer.

Actually setting the opacity on a QPainter level is a dead-end, because it doesn't "flatten" the artwork before applying the opacity change. What we do here is instead draw the source onto a temporary QImage destination, and then modify the opacity of the destination image before painting that onto the final painter destination. It's the same approach used for layer-wide opacity applied to a vector layer.

> I'm not sure about the performance impact though. I also am not sure that more complex effects using gradient of alpha blending will be doable in the full vector solution.

In this situation it's fine to rasterise the output mask layer -- it's unavoidable. But the key is to make sure it's ONLY the mask layer which is being rasterised, and not the rest of the map content.

> QSvgGenerator does not seem to handle clipping at all. Thus selective masking may have to stay fully rasterized for svg export from the layout editor. Any idea about that ?

That's a known issue in Qt, which has been around for the life of QGIS. Clipping paths are being more and more utilised in QGIS, making the svg output option less and less useful. The fix lies in Qt here, so please don't try to work around it on the QGIS side.

> Last small issue is that when the dpi in the layout export tool is set to something different than the fixed DPI of QPicture (96 on my machine), the exported pdf is scaled out of the frame. I think this could be fixed by applying a transformation to the drawn qpicture using a ratio between default and export dpi, but maybe there is a better way ?

This sounds familiar -- if I remember correctly I had to deal with it with the effects framework too. Have a look at QgsPaintEffect::fixQPictureDpi -- that same fix may be required. (it's also used in labeling)

> The code is available here. The full vector implementation lives next to the old rasterized implementation. This is a WIP experimentation probably unstable and not implemented in the most efficient way as I was not familiar at all with this part of QGIS : ) I will continue to work on this to try to fix all those issues.

Thanks for the progress update! Exciting stuff...

**Comment by obrix** [comment]
Small update ! I fixed some of the issues :

* Regarding the opacity feature :

The opacity of the mask is now taken into account. The rasterized mask image is combined with the first rendering pass and integrated into the final result as a QImage. If the mask is totally opaque then the QImage integrated is fully transparent, but if the mask has some transparency then the masked part appears in the same way it did in the full rasterized implementation.

Here is an example (Open in Inkscape to have the best rendering) :

[testdemo_transparency.pdf](https://github.com/qgis/QGIS-Enhancement-Proposals/files/5069222/testdemo_transparency.pdf)

* Regarding the DPI issue :

> This sounds familiar -- if I remember correctly I had to deal with it with the effects framework too. Have a look at QgsPaintEffect::fixQPictureDpi -- that same fix may be required. (it's also used in labeling)

Yes I applied the scaling and it fixed the DPI issue.

* Regarding the phantom lines

@nyalldawson What are the symbols that you see duplicated ? Because the rendering is done in two passes some of the symbols are indeed duplicated, for example the yellow parts of the lines in "test demo" are drawn two times : as part of the second pass that separates unmasked symbols, and as part of the first pass. This should not be an issue. I am not sure the phantom lines are coming from these symbols, but maybe these are not the symbols you were referring to ? Furthermore I opened the pdf in various tools (browsers and dedicated softwares), and it yielded various results (some show the lines, some don't).

* Regarding the SVG issue : 

> That's a known issue in Qt, which has been around for the life of QGIS. Clipping paths are being more and more utilised in QGIS, making the svg output option less and less useful. The fix lies in Qt here, so please don't try to work around it on the QGIS side.

Yep I did not try to fix it :)

* Regarding scoped clipping :

I did not investigate this idea yet. But because of the implementation all the clipping is done in separated painters so it should not interfere with the main painter clip path. @nyalldawson Do you have any idea regarding how I could test that it works properly ?

**Comment by haubourg** [comment]
Hi, can someone with issue management rights assign this issue to @obrix ? 
 (ping @nyalldawson @NathanW2  @andreasneumann ?  )

**Comment by nyalldawson** [comment]
@obrix 

Regarding the phantom lines
> @nyalldawson What are the symbols that you see duplicated ? Because the rendering is done in two passes some of the symbols are indeed duplicated, for example the yellow parts of the lines in "test demo" are drawn two times : as part of the second pass that separates unmasked symbols, and as part of the first pass. This should not be an issue. I am not sure the phantom lines are coming from these symbols, but maybe these are not the symbols you were referring to ? Furthermore I opened the pdf in various tools (browsers and dedicated softwares), and it yielded various results (some show the lines, some don't).

I've attached a screencast showing the issue. 
![Peek 2020-08-18 15-16](https://user-images.githubusercontent.com/1829991/90473239-c5851080-e165-11ea-94c9-b517fb5263e9.gif)

Notice that the yellow lines are duplicated and are present twice in the pdf file.


> Regarding scoped clipping :
I did not investigate this idea yet. But because of the implementation all the clipping is done in separated painters so it should not interfere with the main painter clip path. @nyalldawson Do you have any idea regarding how I could test that it works properly ?

Try testing with the new "clip shape" option set for a layout map, so that you have a non-rectangular map. That would be a good test

**Comment by obrix** [comment]
@nyalldawson 

> Notice that the yellow lines are duplicated and are present twice in the pdf file.

Yes those are the symbols I was referring to. What can be done here is to also clip the yellow part duplicated located "under" the upper layer with an inverted clip path. I tried this but it made the line artefact worse somehow. I can try it again, maybe it can be acceptable.

> Try testing with the new "clip shape" option set for a layout map, so that you have a non-rectangular map. That would be a good test

Thanks I will try that.

**Comment by nyalldawson** [comment]
The duplicate rendering is a serious issue - if the yellow was partially transparent then you'd see the underlying duplicate layer. It'll need addressing before this work is mergable.

**Comment by obrix** [comment]
@nyalldawson Yes I understand. This is an example with the added clipping step for "duplicated" yellow lines : 

[test_clipping_secondpass.pdf](https://github.com/qgis/QGIS-Enhancement-Proposals/files/5149230/test_clipping_secondpass.pdf)

This makes the duplicated yellow path only visible under the mask location (as it should be). But like I said the lines artifact issue is worse (even visible in Inkscape). I don't really understand where this artifact comes from, maybe some precision issues in the clipping process : 

I use a rect painter path the size of the image (Path A) subtracted with the mask painter path which is a set of circle symbols (Path B). Drawing with a clipping path set to A - B and then also drawing with a clipping path set to B gives me the lines you can see at the limits between (A - B) and B.

**Comment by obrix** [comment]
@nyalldawson Small update : 

Regarding the "phantom lines" issue, it seems it simply comes from antialiasing in clipped drawing and not clipping precision issues. I could not treat the issue in the same way I did it in the rasterized version because the composition stage is very different. 

To go around the issue I changed the way the rendering of symbol layers is done for the second rendering pass. In the existing implementation, all symbols were rendered in the first pass, then the second pass was re rendering only the symbols that are not masked. The composition was done from the results of the two passes.

What I did is that instead of re rendering only symbols that are not masked, I re rendered every symbol layers but each one in a separated qpicture. The composition is done from all the resulting qpictures. This allows to clip separately the symbollayers that are masked and then fix several issues that came from the way the composition was done :
* No more phantom lines 
* No more duplicated symbols
* Symbols draw order issues that were present with the way the composition was done are now gone.

This is the result on the same test project :

[test.pdf](https://github.com/qgis/QGIS-Enhancement-Proposals/files/5231017/test.pdf)

The main drawback I see is a heavier second pass regarding rendering time, all symbols are re rendered and not just a subset + the qpainter switch to separate the layers. Also even if the second pass redraw everything separately, the combined result of the first pass is still needed to add the transparency effects to the final composition.

**Comment by troopa81** [comment]
I'm working on a new PR to bring this QEP alive based on previous work https://github.com/qgis/QGIS/pull/38977 https://github.com/qgis/QGIS/pull/43205

I'have already managed to rebase last PR on current master and fix few things. I'm working on the different artefacts reported in previous work and still have to add tests and documentation. I hope I could propose a PR soon.

**Comment by anitagraser** [comment]
Hi @troopa81, have you already sent a final report for this funded enhancement? It would be great if you could add a link in https://github.com/qgis/PSC/issues/43

**Comment by troopa81** [comment]
@anitagraser No report sent yet, because the job is not done. I hope I could finish soon. Do you want me to send a report to say that it's in progress?

**Comment by anitagraser** [comment]
@troopa81 thank you for the update. Good to know that you are on it. It's not necessary to send an interim report but please remember the report when you are done. Thanks.
